Apps/TimelineApp/core/sensor/src/main/java/uk/ac/cam/cares/jps/sensor/SensorService.java

```diff
@@ -317,30 +312,87 @@ private Notification getNotification() {
         return notification;
     }
 
+
+    /**
+     * Checks if activity permission has been granted.
+     * @return a boolean indicating false if permission has not been granted and true if it has.
+     */
+    private boolean checkActivityRecognitionPermission() {
+        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACTIVITY_RECOGNITION)
+                != PackageManager.PERMISSION_GRANTED) {
+            LOGGER.warn("Activity Recognition permission not granted. Unable to start service.");
+            return false; // Permission not granted
+        }
+        return true; // Permission granted
+    }
+
+
     @Override
     public void onDestroy() {
+
+        // cancel timers for workers
+        if (timerB != null) {
+            timerB.cancel();
+        }
+        if (timer != null) {
+            timer.cancel();
+        }
+
         // Unregister the NetworkChangeReceiver when the service is destroyed
         if (networkChangeReceiver != null) {
             unregisterReceiver(networkChangeReceiver);
             networkChangeReceiver = null;
         }
 
-        LOGGER.info("Stopping sensor service");
+        // Stop activity recognition updates
+        if (activityRecognitionClient != null) {
+            PendingIntent pendingIntent = PendingIntent.getService(this, 0,
+                    new Intent(this, ActivityRecognitionService.class), PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE);
+            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACTIVITY_RECOGNITION) != PackageManager.PERMISSION_GRANTED) {
+                // TODO: Consider calling
+                //    ActivityCompat#requestPermissions
+                // here to request the missing permissions, and then overriding
+                //   public void onRequestPermissionsResult(int requestCode, String[] permissions,
+                //                                          int[] grantResults)
+                // to handle the case where the user grants the permission. See the documentation
+                // for ActivityCompat#requestPermissions for more details.
+                return;
+            }
+            activityRecognitionClient.removeActivityUpdates(pendingIntent);
+        }
+
+
+        LOGGER.info("Stopping sensor service 1");
         try {
-            sensorHandlerManager.stopSensors();
+            if (sensorHandlerManager != null) {
```

Comment:
i added this when i was debugging the thread stuff because it gave me the error that it was unable to stop the service - i can remove it but if the manager is null can it call stopsensors? won't it crash